libs/common/src/main/java/org/opensearch/common/crypto/CryptoProvider.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.crypto;
+
+import org.opensearch.common.io.InputStreamContainer;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Crypto provider abstractions for encryption and decryption of data. Allows registering multiple providers
+ * for defining different ways of encrypting or decrypting data.
+ */
+public interface CryptoProvider {
+
+    /**
+     * To initialise or create a new crypto metadata to be used in encryption. This is needed to set the context before
+     * beginning encryption.
+     *
+     * @return crypto metadata instance
+     */
+    Object initEncryptionMetadata();
+
+    /**
+     * To load crypto metadata to be used in encryption from content header.
+     * Note that underlying information in the loaded metadata object is same as present in the object created during
+     * encryption but object type may differ.
+     *
+     * @return crypto metadata instance used in decryption.
+     */
+    Object loadEncryptionMetadata(EncryptedHeaderContentSupplier encryptedHeaderContentSupplier) throws IOException;
```

Comment:
Can we have a better well defined entity rather than `Object`. Why is init a contract, could this be the ctor in the concrete implementations

Reply:
No, it can't be created in ctor as this falls in request path. On each encryption or decryption request a metadata object is created.

If you look at encryption-decryption abstractions, all of those have a mandatory metadata object requirement. This is to ensure that all operations on part or whole of content happen with same metadata. This also allows us to change metadata settings dynamically without affecting ongoing operations. So, before any encrypt op a new metadata object needs to be created by the caller and before any decrypt op the metadata object has to be loaded by the caller.

Now, coming to whether it can be moved to a generic type. I guess not because then caller has to understand the metadata type which only makes sense within a crypto provider. Also, it can be completely different for different crypto providers.

Reply:
> No, it can't be created in ctor as this falls in request path. On each encryption or decryption request a metadata object is created.

Yes. we would need to it on per  object encryption only 



> Can we have a better well defined entity rather than Object. Why is init a contract, could this be the ctor in the concrete implementations

We have now made it generic now .